### PR TITLE
Restore all jobs in the job state map (bugfix)

### PR DESCRIPTION
**Pull request by Hook25**
## Description

This fixes the bug with the session that when a test is de-selected, any resume of the session will make the submission.json generation fail. This was because the job_state_map is not entirely re-loaded after the resume, but only the tests that were selected are correctly put in the map. 

Additionally, the validation of tests would not cover rejected tests (which is dangerous, as a test may be deselected for a reason, and that reason may not hold anymore). 

Finally there was a bug in the session resume where tests were only incidentally not re-run as the rejected job list was not re-evaluated after resume.

Minor: this fixes a bug with `ExpectNot`, as it wouldn't work if the checkbox controller didnt finish running, now it does

## Resolved issues

Fixes: CHECKBOX-1627
Fixes: https://github.com/canonical/checkbox/issues/1557

## Documentation

N/A

## Tests

This adds a new metabox scenario for the regression

### CODE REVIEW

File: checkbox-ng/plainbox/impl/session/test_resume.py [hunk 1]
```diff
@@ -2031,6 +2031,41 @@ def test_simple_session(self):
         # as specific tests for restoring results are written elsewhere
         self.assertEqual(session.job_state_map[job.id].result.outcome, "pass")
 
+    def test_simple_session_with_rejected(self):
+        """
+        verify that _restore_SessionState_jobs_and_results() works when
+        faced with a representation of a simple session (no generated jobs
+        or anything "exotic").
```
Comment (pieqq) [inline checkbox-ng/plainbox/impl/session/test_resume.py:2038]: This is the same comment as `test_simple_session`, I guess you may want to update it to mention the rejected jobs.

File: checkbox-ng/plainbox/impl/session/test_resume.py [hunk 2]
```diff
@@ -2031,6 +2031,41 @@ def test_simple_session(self):
...
+        session = SessionState([job, job1])
+        helper._restore_SessionState_jobs_and_results(session, session_repr)
+        # Session still has one job in it
+        self.assertEqual(session.job_list, [job, job1])
+        # Resources don't have anything (no resource jobs)
+        self.assertEqual(session.resource_map, {})
+        # The result was restored correctly. This is just a smoke test
+        # as specific tests for restoring results are written elsewhere
+        self.assertEqual(session.job_state_map[job.id].result.outcome, "pass")
+        self.assertIn(job1.id, session.job_state_map)
```
Comment (pieqq) [inline checkbox-ng/plainbox/impl/session/test_resume.py:2067]: Add a comment like `# job1 is a rejected job, but it should still be restored in the job_state_map after the session is resumed`.

**Review by pieqq — CHANGES_REQUESTED** [review]
Great stuff! Thanks for spending the time to figure out this issue.

For the unit tests, it looks like `SessionJobsAndResultsResumeTests` is parametrized somehow, so it might be worth making sure the right version of `SessionResumeHelper` is tested.

Otherwise, I really like the Metabox test, it's very clear.

A few other comments in the unit tests, but otherwise I'd say it's good to land!

File: checkbox-ng/plainbox/impl/session/test_resume.py (hunk 1 above)
Comment (Hook25) [reply checkbox-ng/plainbox/impl/session/test_resume.py:2038]: done

File: checkbox-ng/plainbox/impl/session/test_resume.py (hunk 2 above)
Comment (Hook25) [reply checkbox-ng/plainbox/impl/session/test_resume.py:2067]: done

**Review by pieqq — APPROVED** [review]
Awesome! Thanks!

File: metabox/metabox/scenarios/restart/agent_respawn.py [hunk 3]
```diff
@@ -198,3 +199,43 @@ class ResumePreservesRejectedJobsStateMap(Scenario):
...
     ]
+
+
+@tag("resume", "manual", "regression")
+class RemoteResumePreservesRejectedJobsStateMap(Scenario):
+    """
+    Check that the job_state_map survives both manual closure and restarts
+
+    This differs from Local because in remote the controller waits for the
+    agent to come back, we loose the output of the rebooting job and we don't
```
Comment (pieqq) [inline metabox/metabox/scenarios/restart/agent_respawn.py:210]:
```suggestion
    agent to come back, we lose the output of the rebooting job and we don't
```

**Review by pieqq — DISMISSED** [review]
Awesome stuff! I just found a little typo (loose/lose), but otherwise it's good to land.

**Pull request merged**